reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+
+#### Mapper parameter
+
+MyBatis Mapper supports two parameters:
+
+- `select 1 from t where id = #{param1}` will be converted to `select 1 from t where id =?` as a Prepared Statement and be "prepared", and the actual parameter will be used for reuse. You can achieve the best performance when using this parameter with the previously mentioned Prepare connection parameters.
+- `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is "prepared" to different parameters, it may cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.
+
+#### Dynamic SQL Batch
+
+[Dynamic SQL - foreach](http://www.mybatis.org/mybatis-3/dynamic-sql.html#foreach)
+
+To support the automatic rewriting of multiple `INSERT` statements into the form of `insert ... values(...), (...), ...`, in addition to configuring `rewriteBatchedStatements=true` in JDBC as mentioned before, MyBatis can also use dynamic SQL to semi-automatically generate batch inserts. Take the following mapper as an example:
+
+```xml
+<insert id="insertTestBatch" parameterType="java.util.List" fetchSize="1">
+  insert into test
+   (id, v1, v2)
+  values
+  <foreach item="item" index="index" collection="list" separator=",">
+  (
+   #{item.id}, #{item.v1}, #{item.v2}
+  )
+  </foreach>
+  on duplicate key update v2 = v1 + values(v1)
+</insert>
+```
+
+This mapper generates an `insert on duplicate key update` statement. The number of `(?,?,?)` following "values" is determined by the number of passed lists. Its final effect is similar to using `rewriteBatchStatements=true`, which also effectively reduces communication overhead between the client and TiDB.
+
+As mentioned before, you also need to note that the Prepared Statements will not be cached after their maximum length exceeds the value of `prepStmtCacheSqlLimit`.
+
+#### Streaming result
+
+The earlier section introduces how to stream read execution results in JDBC. In addition to the corresponding configurations of JDBC, if you want to read a super large result set in MyBatis, you also need to note that:
+
+- You can set `fetchSize` for a single SQL statement in the mapper configuration (see the previous code block). Its effect is equivalent to calling `setFetchSize` in JDBC.
+- You can use the query interface with `ResultHandler` to avoid getting the entire result set at once.
+- You can use the `Cursor` class for stream reading.
+
+If you configure mappings using XML, you can stream read results by configuring `fetchSize="-2147483648"`(`Integer.MIN_VALUE`) in the mapping's `<select>` section.
+
+```xml
+<select id="getAll" resultMap="postResultMap" fetchSize="-2147483648">
+  select * from post;
+</select>
+```
+
+If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.
```

Comment:
```suggestion
If you configure mappings using code, you can add the `@Options(fetchSize = Integer.MIN_VALUE)` annotation and keep the type of results as `Cursor` so that the SQL results can be read in streaming.
```